cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -277,6 +277,79 @@ public static MetadataEntity ofDataset(String namespace, String datasetName) {
     return subParts;
   }
 
+  /**
+   * @return a {@link String} which describes the {@link MetadataEntity} for a user in plain english.
+   * If the {@link MetadataEntity} represents a known CDAP entity then the description is worded to show relations
+   * in the hierarchy.
+   */
+  public String getDescription() {
+    return getDescription(new StringBuilder(), getType());
+  }
+
+  private String getDescription(StringBuilder builder, String type) {
+    switch (type) {
+      case MetadataEntity.NAMESPACE:
+        builder.append(String.format("%s: %s", MetadataEntity.NAMESPACE, getValue(MetadataEntity.NAMESPACE)));
+        return builder.toString();
+      case MetadataEntity.APPLICATION:
+        String description = String.format("%s: %s", MetadataEntity.APPLICATION, getValue(MetadataEntity.APPLICATION));
+        // MetadataEntity can be created without version and in that case cdap treats it as default version
+        // internally but for no surprises to user no need to display version if user specifically created this
+        // without version. If the cdap system has created the MetadataEntity it will have version info set even for
+        // default version
+        if (containsKey(MetadataEntity.VERSION)) {
```

Comment:
how come we say "app ... deployed in ... ns" only if it contains version key?
Shouldn't we still say "deployed in" otherwise?